### PR TITLE
Upgrade to humble

### PR DESCRIPTION
**Pull request by LiyouZhou**
- Upgrade all packages to humble
- add lifecycle api
- tested cpp examples to work

@mvukov I didn't work on the rclpy package. Would you be able to help?

### CODE REVIEW

**Review by mvukov — COMMENTED** [review]
This is super cool, many thanks for digging into this.

Please take a look at the latest main. I now use official ros2.repos files from github.com/ros2/ros2 to auto-generate http_archive targets for ROS2 packages, see repositories/private/BUILD.bazel. And I also started using patches like you do in this PR :+1: .

I'll definitely help with rclpy, and possibly other stuff along the way. I just need a bit of time to wrap my head around this.

Regarding cyclonedds: I am really considering to use rules_foreign_cc + it's cmake extension here instead of maintaining this complicated BUILD file. WDYT? From what I know, it should work just fine with cross-compilation.

**Comment by mvukov** [comment]
If you have some time, please merge the latest main and resolve merge conflicts.

**Comment by LiyouZhou** [comment]
Merged with main, tested with examples/chatter and examples/actions, all cpp targets work as expected.

**Comment by LiyouZhou** [comment]
Re cyclone: in my experience cmake targets work surprising well, but because this extra layer of dark magic, bit difficult to debug any build issues. I will contribute another patch after this one is merged to introduce iceoryx backend for cyclone and I used cmake target for that.

Interestingly [iceoryx](https://github.com/eclipse-iceoryx/iceoryx) is switching to a bazel build system I wonder if cyclonedds will follow suite. We should raise an issue in their github repo :blush:.

**Comment by LiyouZhou** [comment]
Hi @mvukov do you think this PR can be merged now?

File: examples/chatter/talker.cc [hunk 1]
```diff
@@ -25,12 +25,8 @@
...
  public:
   MinimalPublisher() : Node("minimal_publisher"), count_(0) {
-    int callback_period_ms = 500;
-    declare_parameter("callback_period_ms");
-    try {
-      callback_period_ms = get_parameter("callback_period_ms").as_int();
-    } catch (const rclcpp::exceptions::ParameterNotDeclaredException&) {
-    }
+    declare_parameter("callback_period_ms", 500);
+    auto callback_period_ms = get_parameter("callback_period_ms").as_int();
```
Comment (mvukov) [inline examples/chatter/talker.cc:29]: In foxy this fires an exception if the node is not started with the param defined. Is this different in Huble?

The idea was to have this by default set to 0.5s and set to something much smaller for tests to take less time.

File: repositories/patches/rclcpp-16.0.1.patch [hunk 2]
```diff
@@ -0,0 +1,25 @@
+diff --git a/rclcpp/src/rclcpp/typesupport_helpers.cpp b/rclcpp/src/rclcpp/typesupport_helpers.cpp
```
Comment (mvukov) [inline repositories/patches/rclcpp-16.0.1.patch:1]: Regarding the patch names: there is no need to have version names therein. If we ever want to support multiple ROS2 distros, that is probably going to be done via different git branches.

File: repositories/patches/rclcpp-16.0.1.patch [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+-#include "ament_index_cpp/get_package_prefix.hpp"
+-#include "ament_index_cpp/get_resources.hpp"
+ #include "rcpputils/shared_library.hpp"
+ #include "rcpputils/find_library.hpp"
+ #include "rosidl_runtime_cpp/message_type_support_decl.hpp"
+@@ -49,11 +47,6 @@ std::string get_typesupport_library_path(
+ #endif
+
+   std::string package_prefix;
+-  try {
```
Comment (mvukov) [inline repositories/patches/rclcpp-16.0.1.patch:18]: Very nice :+1:

File: repositories/patches/rmw_implementation_library_path.patch [hunk 4]
```diff
@@ -1,41 +1,58 @@
...
++#else
 +  const std::string library_path = RMW_LIBRARY_PATH;
++  try {
++    return std::make_shared<rcpputils::SharedLibrary>(library_path.c_str());
++  } catch (const std::exception & e) {
++    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING(
++      "failed to load shared library '%s' due to %s",
++      library_path.c_str(), e.what());
++    return nullptr;
++  }
```
Comment (mvukov) [inline repositories/patches/rmw_implementation_library_path.patch:53]: In fact, we just need this part. I don't think we need `#ifndef RMW_LIBRARY_PATH` stuff.

File: repositories/patches/rosidl_rm_unnecessary_asserts.patch [hunk 5]
```diff
@@ -1,72 +1,36 @@
...
      assert input_file.suffix == '.action'
  
-     abs_input_file = package_dir / input_file
--    print(f'Reading input file: {abs_input_file}')
-     abs_input_file = package_dir / input_file
-     content = abs_input_file.read_text(encoding='utf-8')
-     action = parse_action_string(package_name, input_file.stem, content)
- 
-     output_file = output_dir / input_file.with_suffix('.idl').name
-     abs_output_file = output_file.absolute()
```
Comment (mvukov) [inline repositories/patches/rosidl_rm_unnecessary_asserts.patch:20]: It's a bit strange that you had to remove this much code. :)

File: repositories/rcl.BUILD.bazel [hunk 6]
```diff
@@ -55,3 +56,24 @@ ros2_c_library(
         "@ros2_rosidl//:rosidl_runtime_c",
     ],
 )
+
+ros2_c_library(
```
Comment (mvukov) [inline repositories/rcl.BUILD.bazel:60]: AFAICS, this is a new library not strictly necessary for migration to humble. WDYT?

File: repositories/rcl_logging.BUILD.bazel [hunk 7]
```diff
@@ -1,15 +1,27 @@
 """ Builds rcl_logging.
 """
 
-load("@com_github_mvukov_rules_ros2//ros2:cc_defs.bzl", "ros2_cpp_library")
+load("@com_github_mvukov_rules_ros2//ros2:cc_defs.bzl", "CPP_COPTS", "ros2_cpp_library")
+
+ros2_cpp_library(
```
Comment (mvukov) [inline repositories/rcl_logging.BUILD.bazel:6]: ros2_c_library

File: repositories/repositories.bzl [hunk 8]
```diff
@@ -49,18 +49,27 @@ def ros2_repositories():
         http_archive,
         name = "libyaml",
         build_file = "@com_github_mvukov_rules_ros2//repositories:libyaml.BUILD.bazel",
-        sha256 = "2c103fc473e904c6fe7580277f1fa16b6d716e54d5e3f32a8913c4850ae03b3f",
-        strip_prefix = "libyaml-acd6f6f014c25e46363e718381e0b35205df2d83",
-        urls = ["https://github.com/yaml/libyaml/archive/acd6f6f014c25e46363e718381e0b35205df2d83.tar.gz"],
+        sha256 = "fa240dbf262be053f3898006d502d514936c818e422afdcf33921c63bed9bf2e",
+        strip_prefix = "libyaml-0.2.5",
```
Comment (mvukov) [inline repositories/repositories.bzl:53]: The commit that is already used is newer than this release. Can we stick to that one?

File: repositories/repositories.bzl [hunk 9]
```diff
@@ -49,18 +49,27 @@ def ros2_repositories():
...
+        urls = ["https://github.com/yaml/libyaml/archive/0.2.5.tar.gz"],
+    )
+
+    maybe(
```
Comment (mvukov) [inline repositories/repositories.bzl:57]: Please rm this definition, it should be used from the auto-gen ros2_repositories_impl.bzl.

File: ros2/cc_defs.bzl [hunk 10]
```diff
@@ -4,7 +4,7 @@
 load("@rules_cc//cc:defs.bzl", "cc_binary", "cc_library")
 
 C_COPTS = ["-std=c11"]
-CPP_COPTS = ["-std=c++14"]
+CPP_COPTS = ["-std=c++17"]
```
Comment (mvukov) [inline ros2/cc_defs.bzl:7]: :+1:

**Review by mvukov — COMMENTED** [review]
Thanks again for digging into this. It's interesting that C++ part can work with humble and the rest essentially with foxy (the launch and test part).

repositories/ros2_repositories_impl.bzl should be auto-generated, please see the doc: https://github.com/mvukov/rules_ros2/blob/main/docs/ros2_repos.md. One you auto-generate the bzl file, the CI should fail, that's expected as probably the rest of the patches or the targets are obsolete. I'll help soon with fixing the rest.

File: examples/chatter/talker.cc (hunk 1 above)
Comment (LiyouZhou) [reply examples/chatter/talker.cc:29]: The second parameter to declare_parameter is the default value.

File: repositories/rcl.BUILD.bazel (hunk 6 above)
Comment (LiyouZhou) [reply repositories/rcl.BUILD.bazel]: Removed

**Comment by LiyouZhou** [comment]
@mvukov Thank you for you review. I think I have covered all your comments. Let me know if you want anything else changed. I think the CI is not running correctly, rclpy still fails to compile in my local repo.

**Comment by mvukov** [comment]
Thanks very much for your efforts. I'll look into rclpy and other upgrade to make this fully working. I'll also investigate CI issues.

**Comment by mvukov** [comment]
I will continue developments in `feature/upgrade_to_humble` branch on this remote. I can confirm that there was a problem with CI. I changed the cache path and then CI fails.

**Comment by mvukov** [comment]
Closing in favor of #19 .

**Pull request closed without merging**